# COMP5318 Assignment 1: Rice Classification

##### Group number: GROUP 16
##### Student 1 SID: 530257633
##### Student 2 SID: 530160436  
##### Student 3 SID: 530238469
##### Student 4 SID: 530162441

## **1. Data Pre-processing**

In [ ]:
# Import all libraries
import numpy as np
import pandas as pd
from sklearn.impute import SimpleImputer
from sklearn.model_selection import StratifiedKFold, cross_val_score, train_test_split, GridSearchCV
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import accuracy_score, f1_score
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.ensemble import AdaBoostClassifier, GradientBoostingClassifier, RandomForestClassifier

In [ ]:
# Ignore future warnings
from warnings import simplefilter
simplefilter(action='ignore', category=FutureWarning)

In [ ]:
# Load the rice dataset: rice-final2.csv
dataset = pd.read_csv("rice-final2.csv", na_values="?")

feature_columns = dataset.columns[:-1]
target_column = dataset.columns[-1]

X_raw = dataset[feature_columns]
y_raw = dataset[target_column]

In [ ]:
# Pre-process dataset
imputer = SimpleImputer(strategy="mean")
scaler = MinMaxScaler()

X = imputer.fit_transform(X_raw)
X = scaler.fit_transform(X)

class_mapping = {"class1": 0, "class2": 1}
y = y_raw.replace(class_mapping).astype(int).to_numpy()

In [ ]:
# Print first ten rows of pre-processed dataset to 4 decimal places as per assignment spec
# A function is provided to assist

def print_data(X, y, n_rows=10):
    """Takes a numpy data array and target and prints the first ten rows.
    
    Arguments:
        X: numpy array of shape (n_examples, n_features)
        y: numpy array of shape (n_examples)
        n_rows: numpy of rows to print
    """
    for example_num in range(min(n_rows, len(X))):
        for feature in X[example_num]:
            print("{:.4f}".format(feature), end=",")

        if example_num == len(X)-1:
            print(y[example_num],end="")
        else:
            print(y[example_num])

print_data(X, y)


0.4628,0.5406,0.5113,0.4803,0.7380,0.4699,0.1196,1
0.4900,0.5547,0.5266,0.5018,0.7319,0.4926,0.8030,1
0.6109,0.6847,0.6707,0.5409,0.8032,0.6253,0.1185,0
0.6466,0.6930,0.6677,0.5961,0.7601,0.6467,0.2669,0
0.6712,0.6233,0.4755,0.8293,0.3721,0.6803,0.4211,1
0.2634,0.2932,0.2414,0.4127,0.5521,0.2752,0.2825,1
0.8175,0.9501,0.9515,0.5925,0.9245,0.8162,0.0000,0
0.3174,0.3588,0.3601,0.3908,0.6921,0.3261,0.8510,1
0.3130,0.3050,0.2150,0.5189,0.3974,0.3159,0.4570,1
0.5120,0.5237,0.4409,0.6235,0.5460,0.5111,0.3155,1


## **2. Build Classifiers**

- Part 1:  Logistic Regression, Naïve Bayes
- Part 2:  KNN, Decision Tree, Ada Boost, Gradient Boost, Random Forest, SVM

### Part 1: Cross-validation without parameter tuning

In [ ]:
## Setting the 10 fold stratified cross-validation
cvKFold=StratifiedKFold(n_splits=10, shuffle=True, random_state=0)

# The stratified folds from cvKFold should be provided to the classifiers

In [ ]:
# Logistic Regression
logr_clf = LogisticRegression(random_state=0, solver="liblinear")
logr_scores = cross_val_score(logr_clf, X, y, cv=cvKFold, scoring="accuracy")
logr_avg_accuracy = np.mean(logr_scores)

In [ ]:
# Naïve Bayes
nb_clf = GaussianNB()
nb_scores = cross_val_score(nb_clf, X, y, cv=cvKFold, scoring="accuracy")
nb_avg_accuracy = np.mean(nb_scores)

### Part 1 Results


In [ ]:
# Print results for each classifier in part 1 to 4 decimal places here:
print(f"LogR average cross-validation accuracy: {logr_avg_accuracy:.4f}")
print(f"NB average cross-validation accuracy: {nb_avg_accuracy:.4f}")

LogR average cross-validation accuracy: 0.9386
NB average cross-validation accuracy: 0.9264


### Part 2: Cross-validation with parameter tuning

In [ ]:
# KNN 
# parameters you may consider
k = [1, 3, 5, 7]
p = [1, 2]

knn_clf = KNeighborsClassifier()
knn_param_grid = {"n_neighbors": k, "p": p}


In [ ]:
# Decision Tree 
# parameters you may consider
max_depth = [3, 5, 7, 10]
min_samples_split = [2, 5, 10]
min_samples_leaf = [1, 2, 4]

dt_clf = DecisionTreeClassifier(random_state=0)
dt_param_grid = {
    "max_depth": max_depth,
    "min_samples_split": min_samples_split,
    "min_samples_leaf": min_samples_leaf
}


In [ ]:
# Ada Boost
# parameters you may consider
n_estimators = [50, 100, 150]
learning_rate = [0.1, 0.2, 0.3, 0.5]

ada_clf = AdaBoostClassifier(random_state=0)
ada_param_grid = {"n_estimators": n_estimators, "learning_rate": learning_rate}

In [ ]:
# Gradient Boost
# parameters you may consider
max_depth = [1, 3, 5, 7]
n_estimators = [50, 100, 150]
learning_rate = [0.1, 0.2, 0.3, 0.5]

gb_clf = GradientBoostingClassifier(random_state=0)
gb_param_grid = {
    "max_depth": max_depth,
    "n_estimators": n_estimators,
    "learning_rate": learning_rate
}

In [ ]:
# Random Forest
# You should use RandomForestClassifier from sklearn.ensemble with information gain and max_features set to ‘sqrt’.
# parameters you may consider
n_estimators = [10, 30, 60, 100]
max_leaf_nodes = [6, 12]

rf_clf = RandomForestClassifier(
    criterion="entropy", max_features="sqrt", random_state=0
)
rf_param_grid = {"n_estimators": n_estimators, "max_leaf_nodes": max_leaf_nodes}


In [ ]:
# SVM
# parameters you may consider
C = [0.01, 0.1, 1, 5]
gamma = [0.01, 0.1, 1, 10]
# optional
kernel = ["rbf"]

svm_clf = SVC(random_state=0)
svm_param_grid = {"C": C, "gamma": gamma, "kernel": kernel}


### Part 2: Results

In [ ]:
# Perform Grid Search with 10-fold stratified cross-validation (GridSearchCV in sklearn). 
# The stratified folds from cvKFold should be provided to GridSearchCV

# This should include using train_test_split from sklearn.model_selection with stratification and random_state=0
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, stratify=y, random_state=0
)

In [ ]:
# KNN
knn_grid = GridSearchCV(knn_clf, knn_param_grid, cv=cvKFold, scoring="accuracy")
knn_grid.fit(X_train, y_train)
knn_test_pred = knn_grid.best_estimator_.predict(X_test)
knn_test_acc = accuracy_score(y_test, knn_test_pred)

print("KNN best k:", knn_grid.best_params_["n_neighbors"])
print("KNN best p:", knn_grid.best_params_["p"])
print(f"KNN cross-validation accuracy: {knn_grid.best_score_:.4f}")
print(f"KNN test set accuracy: {knn_test_acc:.4f}")

KNN best k: 7
KNN best p: 2
KNN cross-validation accuracy: 0.9375
KNN test set accuracy: 0.9250


In [ ]:
# Decision Tree
dt_grid = GridSearchCV(dt_clf, dt_param_grid, cv=cvKFold, scoring="accuracy")
dt_grid.fit(X_train, y_train)
dt_test_pred = dt_grid.best_estimator_.predict(X_test)
dt_test_acc = accuracy_score(y_test, dt_test_pred)

print("DT best max_depth:", dt_grid.best_params_["max_depth"])
print("DT best min_samples_split:", dt_grid.best_params_["min_samples_split"])
print("DT best min_samples_leaf:", dt_grid.best_params_["min_samples_leaf"])
print(f"DT cross-validation accuracy: {dt_grid.best_score_:.4f}")
print(f"DT test set accuracy: {dt_test_acc:.4f}")

DT best max_depth: 3
DT best min_samples_split: 2
DT best min_samples_leaf: 1
DT cross-validation accuracy: 0.9357
DT test set accuracy: 0.9429


In [ ]:
# Ada Boost
ada_grid = GridSearchCV(ada_clf, ada_param_grid, cv=cvKFold, scoring="accuracy")
ada_grid.fit(X_train, y_train)
ada_test_pred = ada_grid.best_estimator_.predict(X_test)
ada_test_acc = accuracy_score(y_test, ada_test_pred)

print(f"Ada best n_estimators: {ada_grid.best_params_['n_estimators']:.4f}")
print(f"Ada best learning_rate: {ada_grid.best_params_['learning_rate']:.4f}")
print(f"Ada cross-validation accuracy: {ada_grid.best_score_:.4f}")
print(f"Ada test set accuracy: {ada_test_acc:.4f}")

Ada best n_estimators: 150.0000
Ada best learning_rate: 0.2000
Ada cross-validation accuracy: 0.9455
Ada test set accuracy: 0.9429


In [ ]:
# Gradient Boost
gb_grid = GridSearchCV(gb_clf, gb_param_grid, cv=cvKFold, scoring="accuracy")
gb_grid.fit(X_train, y_train)
gb_test_pred = gb_grid.best_estimator_.predict(X_test)
gb_test_acc = accuracy_score(y_test, gb_test_pred)

print(f"GB best max_depth: {gb_grid.best_params_['max_depth']:.4f}")
print(f"GB best n_estimators: {gb_grid.best_params_['n_estimators']:.4f}")
print(f"GB best learning_rate: {gb_grid.best_params_['learning_rate']:.4f}")
print(f"GB cross-validation accuracy: {gb_grid.best_score_:.4f}")
print(f"GB test set accuracy: {gb_test_acc:.4f}")

GB best max_depth: 1.0000
GB best n_estimators: 50.0000
GB best learning_rate: 0.1000
GB cross-validation accuracy: 0.9446
GB test set accuracy: 0.9429


In [ ]:
# Random Forest
rf_grid = GridSearchCV(rf_clf, rf_param_grid, cv=cvKFold, scoring="accuracy")
rf_grid.fit(X_train, y_train)
rf_test_pred = rf_grid.best_estimator_.predict(X_test)
rf_test_acc = accuracy_score(y_test, rf_test_pred)
rf_macro_f1 = f1_score(y_test, rf_test_pred, average="macro")
rf_weighted_f1 = f1_score(y_test, rf_test_pred, average="weighted")

print(f"RF best n_estimators: {rf_grid.best_params_['n_estimators']:.4f}")
print(f"RF best max_leaf_nodes: {rf_grid.best_params_['max_leaf_nodes']:.4f}")
print(f"RF cross-validation accuracy: {rf_grid.best_score_:.4f}")
print(f"RF test set accuracy: {rf_test_acc:.4f}")
print(f"RF test set macro average F1: {rf_macro_f1:.4f}")
print(f"RF test set weighted average F1: {rf_weighted_f1:.4f}")

RF best n_estimators: 30.0000
RF best max_leaf_nodes: 6.0000
RF cross-validation accuracy: 0.9411
RF test set accuracy: 0.9429
RF test set macro average F1: 0.9414
RF test set weighted average F1: 0.9427


In [ ]:
# SVM
svm_grid = GridSearchCV(svm_clf, svm_param_grid, cv=cvKFold, scoring="accuracy")
svm_grid.fit(X_train, y_train)
svm_test_pred = svm_grid.best_estimator_.predict(X_test)
svm_test_acc = accuracy_score(y_test, svm_test_pred)

print(f"SVM best C: {svm_grid.best_params_['C']:.4f}")
print(f"SVM best gamma: {svm_grid.best_params_['gamma']:.4f}")
print("SVM best kernel:", svm_grid.best_params_["kernel"])
print(f"SVM cross-validation accuracy: {svm_grid.best_score_:.4f}")
print(f"SVM test set accuracy: {svm_test_acc:.4f}")

SVM best C: 5.0000
SVM best gamma: 1.0000
SVM best kernel: rbf
SVM cross-validation accuracy: 0.9429
SVM test set accuracy: 0.9321


### Test your code

In [ ]:
# load the test dataset to test out your model
# This section is for testing runnability only.

test_dataset = pd.read_csv("test-before.csv", na_values="?")
test_feature_columns = test_dataset.columns[:-1]
test_target_column = test_dataset.columns[-1]

test_X_raw = test_dataset[test_feature_columns]
test_y_raw = test_dataset[test_target_column]

test_imputer = SimpleImputer(strategy="mean")
test_scaler = MinMaxScaler()
test_X = test_imputer.fit_transform(test_X_raw)
test_X = test_scaler.fit_transform(test_X)
test_y = test_y_raw.replace(class_mapping).astype(int).to_numpy()

print(
    f"Test-before preprocessing succeeded: X shape={test_X.shape}, y shape={test_y.shape}"
)

# Lightweight sanity check model to verify the pipeline runs end-to-end.
test_X_train, test_X_test, test_y_train, test_y_test = train_test_split(
    test_X, test_y, test_size=0.2, stratify=test_y, random_state=0
)

test_logr = LogisticRegression(random_state=0, solver="liblinear")
test_logr.fit(test_X_train, test_y_train)
test_sanity_acc = test_logr.score(test_X_test, test_y_test)
print(f"Test-before sanity check (LogR) accuracy: {test_sanity_acc:.4f}")


Test-before preprocessing succeeded: X shape=(209, 6), y shape=(209,)
Test-before sanity check (LogR) accuracy: 0.5952


## **3. Reflection and Discussion**



The assignment compared the performance of various classification models on the same dataset, including Logistic Regression, Naive Bayes, KNN, Decision Tree, AdaBoost, Gradient Boosting, Random Forest, and SVM. We use 10-fold stratified cross-validation and evaluation on the test set to observed different performance and generalization ability among the different models.

In Part 1, Logistic Regression (0.9386) ‘s performance is slightly better than Naïve Bayes (0.9264), that indicated Logistic Regression as a baseline model has good stability and performance.

After hyperparameter tuning in Part 2, overall performance improved slightly for most models. Gradient Boosting has the best performance with the highest cross-validation accuracy (0.9446) and  the highest test accuracies (0.9429). It demonstrates the optimal generalization ability on this dataset.
Random forest also performed  well with high macro F1 (0.9414) and weighted F1 (0.9427), that indicated it has strong stability and good class balance ability , making it a reliable classification model.
Although the performance in decision Tree shows lower cross-validation accuracy but the test accuracy rate are very  high(0.9429). This indicates that the decision boundary of this dataset is relatively simple, don’t require complex model to achieve good classification results and the model don’t show significant overfitting. AdaBoost performs well in stable learning, but its performance is not better than advanced ensemble models.
KNN and SVM performed relatively poorly. Although KNN had the optimal parameter k=7 but overall performance was still inferior to other models. Maybe because in high-dimensional Spaces the discriminatory ability of distance metrics will decrease. While SVM also had a high cross-validation accuracy (0.9429) but its test accuracy was low (0.9321), showing a weak generalization ability.

Overall, ensemble learning methods performed best especially Gradient Boosting, which outperformed other models in accuracy and generalization ability. In contrast, single models like KNN and SVM performed slightly worse. Hyperparameter tuning has brought certain improvements to the model performance but the the affect is limited,because the baseline model already fit the dataset well, and the dataset structure is relatively simple. However, tuning helps to find better configurations and ensure the stability and reliability of performance for different classifiers. Furthermore, the small difference between the cross-validation accuracy and the test set accuracy of each model that indicated the models have a good generalization ability and no obvious overfitting was oserved.


## **AI Acknowledgement**




This assignment was completed independently by the Group 16. Only use ChatGPT to check understanding of the requirements and check the grammar of the "Reflection And Discussion" section. All analysis, code, and report are original work by Group 16.